File: 46-permutations/46-permutations.scala
import scala.collection.mutable._

object Solution {
  def permute(nums: Array[Int]): List[List[Int]] = {
    val set: Set[Int] = Set.from(nums)
    val res: ListBuffer[List[Int]] = ListBuffer()
    helper(set, ListBuffer(), res)
    res.toList
  }

  def helper(
      nums: Set[Int],
      cur: ListBuffer[Int],
      res: ListBuffer[List[Int]]
  ): Unit = {
    if (nums.isEmpty) {
      res += cur.toList
    } else {
      for (n <- nums) {
        cur += n
        val next = nums.filter(_ != n)
        helper(next, cur, res)
        cur.remove(cur.length - 1)
      }
    }
  }
}

File: 46-permutations/46-permutations.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();
    
    public List<List<Integer>> permute(int[] nums) {
        helper(nums, new LinkedList<>(), new HashSet<>());
        return result;
    }
    
    void helper(int[] nums, LinkedList<Integer> current, Set<Integer> visited) {
        if (visited.size() == nums.length) {
            result.add(List.copyOf(current));
            return;
        }
        
        for(int i = 0; i < nums.length; i++) {
            if (visited.contains(i)) continue;
            
            visited.add(i);
            current.add(nums[i]);
            helper(nums, current, visited);
            current.removeLast();
            visited.remove(i);
        }
    }
}